man/man3/test-tap.3
.\" Automatically generated by Pod::Man 4.10 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is >0, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{\
.    if \nF \{\
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{\
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "STDIN 1"
.TH "test-tap" 3 "2023-01-01" "md2man v0.1.0" "TAP Testing for Bash"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "TEST-TAP-BASH"
.IX Header "TEST-TAP-BASH"
\&\s-1TAP\s0 Testing for Bash
.SS "\s-1SYNOPSIS\s0"
.IX Subsection "SYNOPSIS"
Write a test file like this. Maybe call it \f(CW\*(C`test/test.t\*(C'\fR:
.PP
.Vb 1
\&    #!/usr/bin/env bash
\&
\&    source bpan.bash \-\-
\&    bpan:source test\-tap
\&
\&    plan 10
\&
\&    ok "$(some\-command)" \*(Aqsome\-command is ok\*(Aq
\&
\&    # or:
\&    try some\-command
\&    ok "$rc" \e
\&       \*(Aqsome\-command is ok\*(Aq
\&
\&    try some\-bad\-command \-\-so\-bad
\&    has "$got" "bad bad bad" \e
\&      "some\-bad\-command error msg contained \*(Aqbad bad bad\*(Aq"
\&
\&    pass \*(AqThis will always pass\*(Aq
\&
\&    fail \*(AqThis will always fail\*(Aq
\&
\&    is $(echo foo) \*(Aqfoo\*(Aq \e
\&       \*(Aqfoo is foo\*(Aq
\&
\&    isnt foo bar \e
\&      "%G isn\*(Aqt %W"  # "foo isn\*(Aqt bar"
\&
\&    like food foo \e
\&      \*(Aqfood is like foo\*(Aq
\&
\&    unlike team I \e
\&      "There\*(Aqs no \*(AqI\*(Aq in \*(Aqteam\*(Aq"
\&
\&    diag "A message for stderr"
\&
\&    note "A message for stdout"
\&
\&    output=( $(ls) )
\&    expected=(README lib bin)
\&    array:is output expected \e
\&      "list files"
.Ve
.PP
Run the test with \f(CW\*(C`prove\*(C'\fR like this:
.PP
.Vb 2
\&    prove test/
\&    prove \-v test/test.t
.Ve
.PP
\&\f(CW\*(C`prove\*(C'\fR knows it's Bash from the first line (the shebang), and it just works.
.PP
You may also run the tests with \f(CW\*(C`bpan test ...\*(C'\fR which just calls \f(CW\*(C`prove ...\*(C'\fR.
.PP
\&\s-1BPAN\s0 modules generally have a Makefile where \f(CW\*(C`make test\*(C'\fR calls \f(CW\*(C`bpan test \-v test/\*(C'\fR.
.SS "\s-1DESCRIPTION\s0"
.IX Subsection "DESCRIPTION"
Test::More <https://metacpan.org/pod/Test::More> is the tried and true testing library for Perl. It uses \s-1TAP\s0 (the Test Anything Protocol). This is the same thing for Bash. For the most part it should work exactly the same.
.SS "\s-1TEST TAP API FUNCTIONS\s0"
.IX Subsection "TEST TAP API FUNCTIONS"
In functions with \f(CW\*(C`[<label\*(C'\fR]>, the label is a (optional) string describing the test that gets printed when the tests are run with verbose (\f(CW\*(C`\-v\*(C'\fR) output.
.PP
The special \*(L"variables\*(R" \f(CW%G\fR and \f(CW%W\fR expand to the \f(CW\*(C`got\*(C'\fR and \f(CW\*(C`want\*(C'\fR string values respectively.
.PP
\&\s-1NOTE:\s0 The got and want values may be altered to fit better in a label string.
.IP "\(bu" 4
\&\f(CW\*(C`plan <number\*(C'\fR>
.Sp
Declare the expected number of tests to be run. If used, must be called before other functions. If not used, you must call \f(CW\*(C`done\-testing\*(C'\fR at the end.
.IP "\(bu" 4
\&\f(CW\*(C`plan skip\-all [<reason\*(C'\fR]>
.Sp
Use this version of \f(CW\*(C`plan\*(C'\fR if it is determined that the tests should \s-1NOT\s0 be run for some reason. Possible reasons are things like a missing command or a certain operating system.
.Sp
Can also be called using \f(CW\*(C`skip\-all [<reason\*(C'\fR]>.
.IP "\(bu" 4
\&\f(CW\*(C`done\-testing [<number\*(C'\fR]>
.Sp
Declare that all tests have been run, optionally with the expected number. If used, must be called after all other assertion functions. If not used, you must call \f(CW\*(C`plan\*(C'\fR at the start.
.IP "\(bu" 4
\&\f(CW\*(C`bail\-on\-fail\*(C'\fR
.Sp
Use this function to declare that any subsequent test failures should terminate the rest of the test script.
.IP "\(bu" 4
\&\f(CW\*(C`try <some command \-\-with=options\*(C'\fR>
.Sp
Try will run a command in a safe way and return the combined stdout and stderr in a variable called \f(CW\*(C`got\*(C'\fR. It will also put the status (return) code in a variable called \f(CW\*(C`rc\*(C'\fR. This is very useful for testing error messages without having the test script error out.
.IP "\(bu" 4
\&\f(CW\*(C`pass [<label\*(C'\fR]>
.Sp
An explicit passing test.
.IP "\(bu" 4
\&\f(CW\*(C`fail [<label\*(C'\fR]>
.Sp
An explicit failing test.
.IP "\(bu" 4
\&\f(CW\*(C`ok "$(<cmd\*(C'\fR)" [<label>]>
.Sp
Pass if cmd runs with a 0 (zero) exit status. You may also call \f(CW\*(C`ok\*(C'\fR with a status code (0\-255) instead of a command. Commands should be formatted like: \f(CW"$( command args... )"\fR; that is wrapped in \f(CW"$(...)"\fR (with actual double quotes).
.IP "\(bu" 4
\&\f(CW\*(C`not\-ok "$(<cmd\*(C'\fR)" [<label>]>
.Sp
Pass if cmd runs with a non-zero exit status. Same rules as \f(CW\*(C`ok\*(C'\fR, but meaning is opposite.
.IP "\(bu" 4
\&\f(CW\*(C`is <got\*(C'\fR <want> [<label>]>
.Sp
Pass if got is equal to want. If want is a multiline string, print a unified diff on failure.
.IP "\(bu" 4
\&\f(CW\*(C`isnt <got\*(C'\fR <unwanted> [<label>]>
.Sp
Pass if got is \s-1NOT\s0 equal to want. Opposite of \f(CW\*(C`is\*(C'\fR.
.IP "\(bu" 4
\&\f(CW\*(C`like <got\*(C'\fR <regex> [<label>]>
.Sp
Pass if got matches regex (bash regular expression).
.IP "\(bu" 4
\&\f(CW\*(C`unlike <got\*(C'\fR <regex> [<label>]>
.Sp
Pass if got doesn't match regex (bash regular expression). Opposite of \f(CW\*(C`like\*(C'\fR.
.IP "\(bu" 4
\&\f(CW\*(C`has <got\*(C'\fR <want> [<label>]>
.Sp
Pass if got contains the want string.
.IP "\(bu" 4
\&\f(CW\*(C`hasnt <got\*(C'\fR <want> [<label>]>
.Sp
Pass if got does \s-1NOT\s0 contain the want string. Opposite of \f(CW\*(C`has\*(C'\fR.
.IP "\(bu" 4
\&\f(CW\*(C`array:is <got\-array\-var\-name\*(C'\fR <want\-array\-var\-name> [<label>]>
.Sp
Pass if the got array has content equal to the want array.
.IP "\(bu" 4
\&\f(CW\*(C`diag <msg\*(C'\fR>
.Sp
Print a message to stderr in a way that the harness knows how to process.
.IP "\(bu" 4
\&\f(CW\*(C`note <msg\*(C'\fR>
.Sp
Print a message to stdout in a way that the harness knows how to process.
.IP "\(bu" 4
\&\f(CW\*(C`bail\-out [<reason\*(C'\fR]>
.Sp
Stop the test script in a way that the harness knows how to process.
.PP
\fIFile System Test Functions\fR
.IX Subsection "File System Test Functions"
.PP
Functions called \f(CW\*(C`ok\-?\*(C'\fR are really just calling \f(CW\*(C`[[ \-? ... ]]\*(C'\fR. For example \f(CW\*(C`ok\-f /path/to/file\*(C'\fR calls \f(CW\*(C`[[ \-f /path/to/file ]]\*(C'\fR.
.PP
You can use a label for the following test funtions, but they will create appropriate labels for you if you don't. Therefore best practice is to not use labels when calling the following functions.
.IP "\(bu" 4
\&\f(CW\*(C`ok\-d <path\*(C'\fR>
.Sp
Pass if path is an existing directory (or a link to a directory).
.IP "\(bu" 4
\&\f(CW\*(C`ok\-e <path\*(C'\fR>
.Sp
Pass if path exists.
.IP "\(bu" 4
\&\f(CW\*(C`ok\-not\-e <path\*(C'\fR>
.Sp
Pass if path does \s-1NOT\s0 exist.
.IP "\(bu" 4
\&\f(CW\*(C`ok\-empty <path\*(C'\fR>
.Sp
Pass if path is an empty file or directory.
.IP "\(bu" 4
\&\f(CW\*(C`ok\-f <path\*(C'\fR>
.Sp
Pass if path is a file (or a link to a file).
.IP "\(bu" 4
\&\f(CW\*(C`ok\-h <path\*(C'\fR>
.Sp
Pass if path is a symlink.
.IP "\(bu" 4
\&\f(CW\*(C`ok\-s <path\*(C'\fR>
.Sp
Pass if path is a non-empty file.
.IP "\(bu" 4
\&\f(CW\*(C`ok\-w <path\*(C'\fR>
.Sp
Pass if path is writable.
.IP "\(bu" 4
\&\f(CW\*(C`ok\-x <path\*(C'\fR>
.Sp
Pass if path is executable.
.SS "\s-1AUTHOR\s0"
.IX Subsection "AUTHOR"
Ingy do\*:t Net ingy@ingy.net <mailto:ingy@ingy.net>
.SS "\s-1COPYRIGHT & LICENSE\s0"
.IX Subsection "COPYRIGHT & LICENSE"
Copyright 2013\-2022 \- Ingy do\*:t Net
.PP
The \s-1MIT\s0 License (\s-1MIT\s0)


Changes
[version "0.1.22"]
date = 2023-01-01T13:10:28Z
done = Update .bpan/config

[version "0.1.21"]
date = 2022-12-15T19:32:04Z
done = Add type to .bpan/config

[version "0.1.20"]
date = 2022-12-13-13-01-56
done = Config changes

[version "0.1.19"]
date = Mon Dec  5 12:51:15 PM PST 2022
done = Fix config

[version "0.1.18"]
date = Mon Dec  5 12:40:47 PM PST 2022
done = Change package.repo to package.source
done = Move Meta to .bpan/config

[version "0.1.17"]
date = Sat Nov 26 02:22:09 PM PST 2022
done = Add test for testing in subprocess

[version "0.1.16"]
date = Thu Nov 17 10:27:59 PM PST 2022
done = Update config to new format

[version "0.1.15"]
date = Fri Nov 11 06:44:55 AM PST 2022
done = Avoid unnecessary subshell functions

[version "0.1.14"]
date = Mon Nov  7 13:02:47 PST 2022
done = Allow source with --

[version "0.1.13"]
date = Sun Oct 30 11:22:25 AM PDT 2022
done = Update BPAN assets
done = Allow color in messages

[version "0.1.12"]
date = Mon Oct 10 02:14:03 PM PDT 2022
done = Fix bugs in label handling

[version "0.1.11"]
date = Fri Sep 30 01:27:28 PM PDT 2022
done = Fix tests for cmp-array rename to array:is
done = Change 'update.template' to 'update.file' in config

[version "0.1.10"]
date = Sun Sep 25 11:50:10 AM EDT 2022
done = Complete the documentation

[version "0.1.9"]
date = Sat Sep 24 09:29:49 AM EDT 2022
done = Update BPAN files
done = Renumber tests
done = Add test functions: ok-empty, ok-h and ok-s

[version "0.1.8"]
date = Fri Sep 23 07:19:46 PM EDT 2022
done = Fix bug with `is` when no label
done = Fix tests to use local lib/test-tap.bash

[version "0.1.7"]
date = Fri Sep 23 02:04:49 PM EDT 2022
done = Refactor to new config file schema
done = Update BPAN assets
done = Fix broken symlink

[version "0.1.6"]
date = Tue Sep 13 01:11:54 PM EDT 2022
done = Add hasnt support

[version "0.1.5"]
date = Sun Sep 11 12:34:49 PM EDT 2022
done = Fix bpan path because bpan changed it

[version "0.1.4"]
date = Sun Sep 11 11:25:37 AM EDT 2022
done = Rename from test-more-bash to test-tap-bash

[version "0.1.3"]
date = Sat Sep 10 02:20:52 PM EDT 2022
done = Add `try` and `has`

[version "0.1.2"]
date = Sun 28 Aug 2022 02:08:32 PM PDT
done = Add file testing functions

[version "0.1.1"]
date = Fri 26 Aug 2022 08:14:15 AM PDT
done = Get tests to all pass
done = Refactor out old test-tap code

[version "0.1.0"]
date = Fri 12 Aug 2022 03:51:21 PM PDT
done = Initial release


lib/test-tap.bash
# test-tap.bash - Complete TAP test framework for Bash
#
# Copyright (c) 2013-2022. Ingy döt Net.

test-tap:main() {
  _test_tap__call_stack_level=1

  [[ ${BASH_SOURCE[0]} ]] ||
    test-tap:die "Error: test-tap-bash must be run under Bash only"
  _test_tap__plan=0
  _test_tap__run=0
  _test_tap__failed=0
  _test_tap__pid=${BASHPID:-0}

  if [[ $# -gt 0 ]]; then
    if ! [[ $# -eq 1 && $1 == -- ]]; then
      if [[ $# -eq 2 && $1 == tests && $2 =~ ^[0-9]+$ ]]; then
        plan "$@"
      else
        test-tap:die 'Usage: source test-tap.bash tests <number>'
      fi
    fi
  fi

  trap test-tap:end EXIT
}

test-tap:label() {
  local label G W
  label=$1
  G=${got-}
  W=${want-}
  G=${G:0:40}
  W=${W:0:40}
  G=${G//$'\n'/␤}
  W=${W//$'\n'/␤}

  label=${label//\%G/$G}
  label=${label//\%W/$W}

  echo -e "$label"
}

test-tap:end() {
  local rc=$?
  test-tap:_check-pid
  if [[ $rc -ne 0 ]]; then
    if [[ ${_test_tap__bail_msg-} ]] ||
       [[ ${test_tap_bail_on_fail-} ]]; then
      local bail=${_test_tap__bail_msg:-"Bailing out on status=$rc"}
      echo "Bail out!  $bail"
      exit $rc
    fi
  fi

  if [[ $_test_tap__plan -eq 0 ]]; then
    if [[ $_test_tap__run -gt 0 ]]; then
      echo "# Tests were run but no plan was declared." >&2
    fi
  else
    if [[ $_test_tap__run -eq 0 ]]; then
      echo "# No tests run!" >&2
    elif [[ $_test_tap__run -ne $_test_tap__plan ]]; then
      local msg="# Looks like you planned $_test_tap__plan tests but ran $_test_tap__run."
      [[ $_test_tap__plan -eq 1 ]] && msg=${msg/tests/test}
      echo "$msg" >&2
    fi
  fi

  local exit_code=0
  if [[ $_test_tap__failed -gt 0 ]]; then
    exit_code=$_test_tap__failed
    [[ $exit_code -gt 254 ]] && exit_code=254
    local msg="# Looks like you failed $_test_tap__failed tests of $_test_tap__run run."
    [[ $_test_tap__failed -eq 1 ]] && msg=${msg/tests/test}
    echo "$msg" >&2
  fi

  exit $exit_code
}

test-tap:_check-pid() {
  if [[ ${BASHPID:-0} -ne ${_test_tap__pid:-0} ]]; then
    test-tap:die "Error: Called test-tap method from a subprocess" 3
  fi
}

test-tap:is-fail() {
  local _test_tap__call_stack_level=
  _test_tap__call_stack_level=$(( _test_tap__call_stack_level + 1 ))
  if [[ $want =~ $'\n' ]]; then
    echo "$got" > /tmp/got-$$
    echo "$want" > /tmp/want-$$
    diff -u /tmp/{want,got}-$$ >&2 || true
    wc /tmp/{want,got}-$$ >&2
    rm -f /tmp/{got,want}-$$
  else
    diag "\
    got: '$got'
  expected: '$want'"
  fi
}

test-tap:isnt-fail() {
    diag "\
      got: '$got'
   expected: anything else"
}

test-tap:like-fail() {
    diag "Got: '$got'"
}

test-tap:unlike-fail() {
    diag "Got: '$got'"
}

test-tap:die() {
  echo "$1" >&2
  exit 1
}

#------------------------------------------------------------------------------
# API Functions (in order of documentation)
#------------------------------------------------------------------------------
plan() {
  test-tap:_check-pid
  if [[ $# -eq 2 && $1 == skip-all ]]; then
    printf "1..0 # SKIP %s\n" "$2"
    exit 0
  fi

  if [[ $# -eq 2 && $1 == tests ]]; then
    shift
  fi

  if [[ $# -eq 1 ]]; then
    [[ $1 =~ ^-?[0-9]+$ ]] ||
      test-tap:die 'Plan must be a number'
    [[ $1 -gt 0 ]] ||
      test-tap:die 'Plan must be greater then 0'
    _test_tap__plan=$1
    printf "1..%d\n" "$_test_tap__plan"
  else
    test-tap:die 'Usage: plan <number>'
  fi
}

skip-all() {
  plan skip-all "$@"
}

done-testing() {
  test-tap:_check-pid
  _test_tap__plan=$_test_tap__run
  echo 1.."${1:-$_test_tap__run}"
}

bail-on-fail() {
  test-tap:_check-pid
  test_tap_bail_on_fail=1
}

try() {
  local keep=''
  if [[ $1 =~ ^--keep=(.*) ]]; then
    keep=${BASH_REMATCH[1]}
    shift
  fi

  cmd="$*"
  set --
  set +e
  got=$(eval "$cmd" 2>&1)
  rc=$?
  set -e

  if [[ $keep ]]; then
    eval "$(grep "^$keep=" <<<"$got" || true)"
    got=$(grep -v "^$keep=" <<<"$got" || true)
  fi
}

pass() {
  test-tap:_check-pid
  echo -e "ok $((++_test_tap__run))${1:+" - $1"}"
}

fail() {
  test-tap:_check-pid
  ((++_test_tap__run))
  IFS=' ' read -r -a c <<<"$(caller $_test_tap__call_stack_level)"
  local file=${c[2]-}
  local line=${c[0]-}
  local label=${1-} callback=${2-}
  echo -e "not ok $_test_tap__run${label:+" - $label"}"
  label=${label:+"'$label'\n#   at $file line $line."}
  label=${label:-"at $file line $line."}
  echo -e "#   Failed test $label" >&2

  [[ $callback ]] && $callback

  local rc=${test_tap_bail_on_fail:-0}
  [[ $rc -eq 0 ]] || exit "$rc"
}

ok() {
  if (exit "${1:-$?}"); then
    pass "${2-}"
  else
    fail "${2-}"
  fi
}

not-ok() {
  if (exit "${1:-$?}"); then
    fail "${2-}"
  else
    pass "${2-}"
  fi
}

is() {
  local got=$1 want=$2 label=${3-}
  label=$(test-tap:label "$label")
  if [[ $got == "$want" ]]; then
    pass "$label"
  else
    fail "$label" test-tap:is-fail
  fi
}

isnt() {
  local _test_tap__call_stack_level=
  _test_tap__call_stack_level=$(( _test_tap__call_stack_level + 1 ))
  local got=$1 want=$2 dontwant=$2 label=${3-}
  label=$(test-tap:label "$label")
  if [[ $got != "$dontwant" ]]; then
    pass "$label"
  else
    fail "$label" test-tap:isnt-fail
  fi
}

like() {
  local got=$1 want=$2 regex=$2 label=${3-}
  label=$(test-tap:label "$label")
  if [[ $got =~ $regex ]]; then
    pass "$label"
  else
    fail "$label" test-tap:like-fail
  fi
}

unlike() {
  local got=$1 want=$2 regex=$2 label=${3-}
  label=$(test-tap:label "$label")
  if [[ ! $got =~ $regex ]]; then
    pass "$label"
  else
    fail "$label" test-tap:unlike-fail
  fi
}

has() {
  local got=$1 want=$2 label=${3-}
  label=$(test-tap:label "$label")
  local n=$'\n'

  if [[ $got == *"$want"* ]]; then
    pass "$label"
  else
    fail "$label"
    if [[ $got == *"$n"* ]]; then
      got="'$n$got$n'"
    else
      got="'$got'"
    fi
    diag "Text '$want' NOT found in: $got"
  fi
}

hasnt() {
  local got=$1 want=$2 label=${3-}
  label=$(test-tap:label "$label")
  local n=$'\n'

  if [[ $got != *"$want"* ]]; then
    pass "$label"
  else
    fail "$label"
    if [[ $got == *"$n"* ]]; then
      got="'$n$got$n'"
    else
      got="'$got'"
    fi
    diag "Text '$want' WAS found in: $got"
  fi
}

array:is() {
  local arrayname="$1[@]"
  local expname="$2[@]"
  local label=${3-}

  local array=("${!arrayname}")
  local expected=("${!expname}")

  is "$(printf "%s\n" "${array[@]}")" \
    "$(printf "%s\n" "${expected[@]}")" \
    "$label"
}

diag() {
  test-tap:_check-pid
  local msg=$*
  msg="# ${msg//$'\n'/$'\n'\# }"
  echo -e "$msg" >&2
}

note() {
  test-tap:_check-pid
  local msg=$*
  msg="# ${msg//$'\n'/$'\n'\# }"
  echo -e "$msg"
}

bail-out() {
  test-tap:_check-pid
  _test_tap__bail_msg=$*
  : "${_test_tap__bail_msg:=No reason given.}"
  exit 255
}


#------------------------------------------------------------------------------
# File System Testing Functions
#------------------------------------------------------------------------------
ok-d() {
  local dir=${1:?}
  local msg=${2:-"'${dir#*//}' is a directory"}
  ok "$([[ -d $dir ]])" "$msg"
}

ok-e() {
  local path=${1:?}
  local msg=${2:-"'${path#*//}' exists"}
  ok "$([[ -e $path ]])" "$msg"
}

ok-empty() {
  local path=${1:?}
  local msg
  if [[ -f $path ]]; then
    msg=${2:-"'${path#*//}' is an empty file"}
    ok "$([[ ! -s $path ]])" "$msg"
  elif [[ -d $path ]]; then
    msg=${2:-"'${path#*//}' is an empty directory"}
    ok "$(shopt -s nullglob; ! [[ $(echo "$path"/*) ]])" "$msg"
  else
    fail "'${path#*//}' does not exist"
  fi
}

ok-not-e() {
  local path=${1:?}
  local msg=${2:-"'${path#*//}' does not exist"}
  ok "$([[ ! -e $path ]])" "$msg"
}

ok-f() {
  local file=${1:?}
  local msg=${2:-"'${file#*//}' is a file"}
  ok "$([[ -f $file ]])" "$msg"
}

ok-h() {
  local link=${1:?}
  local msg=${2:-"'${link#*//}' is a symlink"}
  ok "$([[ -h $link ]])" "$msg"
}

ok-s() {
  local file=${1:?}
  local msg=${2:-"'${file#*//}' is a non-empty file"}
  ok "$([[ -s $file ]])" "$msg"
}

ok-w() {
  local path=${1:?}
  local msg=${2:-"'${path#*//}' is writable"}
  ok "$([[ -w $path ]])" "$msg"
}

ok-x() {
  local path=${1:?}
  local msg=${2:-"'${path#*//}' is executable"}
  ok "$([[ -x $path ]])" "$msg"
}

test-tap:main "$@"
